Mathlib/Data/Matroid/Rank/Finite.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2025 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import Mathlib.Data.Matroid.Closure
+
+/-!
+# Finite-rank sets
+
+For `M : Matroid α` and `X : Set α`,
+`Matroid.RkFin M X`  means that every basis of `X` in `M` is finite,
+or equivalently that the restriction to `X` is `Matroid.RankFinite`.
+`RkFin` sets in a matroid are the largest class of sets for which one can do nontrivial
+integer arithmetic involving the rank function.
+
+# Implementation Details
+
+Unlike most set predicates on matroids, a set `X` with `M.RkFin X` need not satisfy `X ⊆ M.E`,
+so may contain junk elements. This seems to be what makes the definition easiest to use.
+-/
+
+variable {α : Type*} {M : Matroid α} {X Y I : Set α} {e : α}
+
+open Set
+
+namespace Matroid
+
+/-- a `RkFin` set in `M` is one whose bases are all finite. -/
+def RkFin (M : Matroid α) (X : Set α) : Prop := (M ↾ X).RankFinite
+
+lemma RkFin.rankFinite (hX : M.RkFin X) : (M ↾ X).RankFinite :=
+  hX
+
+/-- A `Basis'` of a `RkFin` set is finite. -/
+lemma RkFin.finite_of_basis' (h : M.RkFin X) (hI : M.Basis' I X) : I.Finite :=
+  have := h.rankFinite
+  (base_restrict_iff'.2 hI).finite
+
+lemma RkFin.finite_of_basis (h : M.RkFin X) (hI : M.Basis I X) : I.Finite :=
+  h.finite_of_basis' hI.basis'
+
+lemma Basis'.finite_of_rkFin (hI : M.Basis' I X) (h : M.RkFin X) : I.Finite :=
+  h.finite_of_basis' hI
+
+lemma Basis.finite_of_rkFin (hI : M.Basis I X) (h : M.RkFin X) : I.Finite :=
+  h.finite_of_basis hI
+
+lemma Basis'.rkFin_of_finite (hI : M.Basis' I X) (hIfin : I.Finite) : M.RkFin X :=
+  ⟨I, hI, hIfin⟩
+
+lemma Basis.rkFin_of_finite (hI : M.Basis I X) (hIfin : I.Finite) : M.RkFin X :=
+  ⟨I, hI.basis', hIfin⟩
+
+lemma Basis'.finite_iff_rkFin (hI : M.Basis' I X) : I.Finite ↔ M.RkFin X :=
+  ⟨hI.rkFin_of_finite, fun h ↦ h.finite_of_basis' hI⟩
+
+lemma Basis.finite_iff_rkFin (hI : M.Basis I X) : I.Finite ↔ M.RkFin X :=
+  hI.basis'.finite_iff_rkFin
```

Comment:
It seems cleaner to me to give these two first, and have the others as aliases. That way, it's also clearer from reading the file which are the ones with mathematical meaning, and which are present just for nice dot notation.